text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+- Feature Name: `generic_integers`
+- Start Date: 2024-08-07
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Adds the builtin types `u<N>` and `i<N>`, allowing integers with an arbitrary size in bits.
+
+# Motivation
+[motivation]: #motivation
+
+## Generalising code for integers
+
+Right now, there's a *lot* of boilerplate for implementing methods for integer primitives. The standard library itself is a great example; almost the entirety of `core::num` uses some gnarly macros to define all sorts of traits for all the
+integer types. One example is `Shl` and `Shr`, which are defined for not just every integer type, but every *combination* of integer types. We could easily do this with const generics instead:
+
+```rust
+impl<const N: usize, const M: usize> Shl<u<M>> for u<N> {
+    type Output = u<N>;
+    #[inline]
+    fn shl(self, rhs: u<M>) -> u<N> {
+        // implementation
+    }
+}
+```
+
+This will decrease compilation time across the entire Rust ecosystem, maybe not by a noticeable amount, but by some amount, due to the presence of macro-based trait implementations across the board.
+
+## Documentation decluttering
+
+Having generic impls would drastically reduce the noise in the "implementations" section of rustdoc. For example, the number of implementations for `Add` for integer types really drowns out the fact that it's also implemented for strings and `std::time` types, which is useful to know too.
+
+## Enum optimisations
+
+Using a smaller number of bits for a value also has the benefit of niche enum optimisations. For example, `u<7>` represents a single ASCII character, and `Option<u<7>>` can be stored in a single byte. Additionally, `Result<u<7>, E>` also takes one byte if `E` is zero-sized.
```

Comment:
This is not an argument, because this is already handled by pattern types ( https://github.com/rust-lang/rust/pull/120131 )

Reply:
These two proposals are not mutually exclusive. Both can coexist.